Properly escape strings in MySQL statement values

devenbansod opened the pull request:
Before submitting pull request, please check that every commit:
- [x] Has proper Signed-Off-By
- [x] Has commit message which describes it
- [x] Is needed on it's own, if you have just minor fixes to previous commits, you can squash them
- [ ] Any new functionality is covered by tests

Use *_real_escape string functions provided by connectors to escape strings while exporting

Fix #12453

Signed-off-by: Deven Bansod devenbansod.bits@gmail.com

nijel commented on libraries/dbi/DBIDummy.php:
```diff
@@ -1256,4 +1256,17 @@ public function fieldFlags($result, $i)
+     */
+    public function escapeString($link, $str)
+    {
+        return '';
```
IMHO it return either original or MySQL quoted string here, otherwise it's pretty confusing.

nijel commented on libraries/DatabaseInterface.php:
```diff
@@ -2649,6 +2649,19 @@ public function fieldFlags($result, $i)
+     */
+    public function escapeString($link, $str)
+    {
+        return $this->_extension->escapeString($link, $str);
```
I'd prefer to have link optional and use $link = $this->getLink($link);. It's really not needed in most cases...

devenbansod replied:
Changes done in the latest commit.

nijel commented on libraries/plugins/export/ExportSql.php:
```diff
@@ -2411,10 +2411,13 @@ public function exportData(
+                                $search,
+                                $replace,
+                                Util::sqlAddSlashes($row[$j])
+                            )
```
Is it intentional to use escapeString and Util::sqlAddSlashes? IMHO it will produce double escaped slashes.

devenbansod replied:
Indeed it would have been wrong. I have removed the extra call in the new commit.

nijel reviewed (changes requested):
Overall I'm not really sure how newly introduced DBI::escapeString relates to Util::sqlAddSlashes. They both IMHO do pretty much same thing.

devenbansod commented:
I feel that `sqlAddSlashes` does not take care of double quotes inside a SQL statement properly, while use of this inbuilt function(s) would do that.

nijel commented:
So probably all uses of sqlAddSlashes should be turned into this new function and sqlAddSlashes can be removed?

devenbansod commented:
I think so but I would welcome a second opinion here. 

The page [here] (http://php.net/manual/en/mysqli.real-escape-string.php) says it encodes the characters `NUL (ASCII 0)`, `\n`, `\r`, `\`, `'`, `"` and the backtick, while `sqlAddSlashes` is just meant to add slashes before the `'` and `\` characters and we also provide an option to encode `\r` and `\n`.

One of the [comment](http://php.net/manual/en/mysqli.real-escape-string.php#46339) on the same page provides some more info.

Also, do you think moving finally to usage of prepared statements might help us in tackling all this issues (as reported by [this person] (http://php.net/manual/en/mysqli.real-escape-string.php#102639) )?

nijel commented:
Using prepared statements would be way better than constructing the SQL queries manually as we do right now. AFAIK it would mean dropping support for the mysql extension, but that really should not matter these days...

As for escaping, the null byte probably should be escaped as well, but as it won't appear in the user input, it's not properly handled by our code. For the newlines I think it should be there as well automatically. It was introduced by 280336063.

devenbansod commented:
Would that mean, for the time till we move completely to prepared statements, we can use this instead of Util::sqlAddSlashes ? 
Also, if we decide to replace all sqlAddSlashes calls with this inbuilt-function, would it be okay to do it in QA_4_6 (like would it be a major change that may break some other thing ?) or should we include this small change for fixing the export bug in QA_4_6 and then target master for the complete replacement ?

nijel commented:
I think using mysql(i)_real_escape_string instead of sqlAddSlashes is correct approach (until we use prepared statements). I'm fine with doing this change in QA_4_6.

devenbansod commented:
Okay. Thanks for the comments. 
I will go ahead and make the changes across other places and would let you know once I feel the PR is ready.

devenbansod commented:
Changes on the way. 
Many tests will fail for now, as I would have to add the mock and stubs with the Mock object of `dbi`.

devenbansod committed:
Properly escape strings in MySQL statement values

Use *_real_escape string functions provided by connectors to escape strings while exporting

Fix #12453

Signed-off-by: Deven Bansod <devenbansod.bits@gmail.com>

Conflicts:
	libraries/server_privileges.lib.php

devenbansod committed:
Fix tests for Table, ThemeManager, Tracker classes

Signed-off-by: Deven Bansod <devenbansod.bits@gmail.com>

devenbansod committed:
Fix tests in Navigation, TableSearchController

Signed-off-by: Deven Bansod <devenbansod.bits@gmail.com>

devenbansod committed:
Fix Navigation tests and Export tests

Signed-off-by: Deven Bansod <devenbansod.bits@gmail.com>

devenbansod committed:
Fixed regressive change in phpunit config file

Signed-off-by: Deven Bansod <devenbansod.bits@gmail.com>

devenbansod committed:
Fixed some tests in Export, Import and Node

Signed-off-by: Deven Bansod <devenbansod.bits@gmail.com>

devenbansod committed:
Fixed test errors in PMD_Common tests

Signed-off-by: Deven Bansod <devenbansod.bits@gmail.com>

devenbansod committed:
Fixed tests in Central Columns

Signed-off-by: Deven Bansod <devenbansod.bits@gmail.com>

devenbansod committed:
Fixed tests in Designer, Insert edit for, UserGroups, User preferences

Signed-off-by: Deven Bansod <devenbansod.bits@gmail.com>

devenbansod committed:
Fix errors with TrackerTest

Signed-off-by: Deven Bansod <devenbansod.bits@gmail.com>

devenbansod committed:
Fixed small error in XML test

Signed-off-by: Deven Bansod <devenbansod.bits@gmail.com>

devenbansod committed:
Fixed some more test errors

Signed-off-by: Deven Bansod <devenbansod.bits@gmail.com>

devenbansod committed:
Fix more test errors

Signed-off-by: Deven Bansod <devenbansod.bits@gmail.com>

devenbansod committed:
Fix more tests

Signed-off-by: Deven Bansod <devenbansod.bits@gmail.com>

devenbansod committed:
clear some more tests

Signed-off-by: Deven Bansod <devenbansod.bits@gmail.com>

devenbansod committed:
Remove tests against unsupported versions
Remove duplicate test, fix other tests

Signed-off-by: Deven Bansod <devenbansod.bits@gmail.com>

devenbansod commented:
Finally the tests are adjusted. :-) 
I think now it's ready for a review.

nijel commented:
I think it looks good, there are still some things to fix:

* There is still one usage left in DatabseInterface::getTablesFull (called by string)
* Another ones in libraries/rte/rte_list.lib.php
* I guess test/libraries/common/PMA_quoting_slashing_test.php should be migrated to test for this new code
* Why not remove sqlAddSlashes code?

devenbansod committed:
Remove call to old methods, fix tests

Remove definition of old method. Remove related tests

Signed-off-by: Deven Bansod <devenbansod.bits@gmail.com>

devenbansod commented:
I have removed the old code (old tests for now as well) and fixed the remaining calls.

But I am not really sure how to cover the `escapeString` function in the tests, since we would have to again add a mock test-double for `GLOBALS['dbi']` to handle the `escapeString` call to the databaseInterface and define the escaped strings that it should return (using may be a `returnValueMap`). 

Any help would be appreciated.

nijel commented:
Hmm, it's true that it's now external code and hard to test...

nijel commented:
Okay, merging now. It would be great if you could review it in master, there were quite some fixes needed.
